fixed-hash/src/hash.rs

```diff
@@ -6,735 +6,646 @@
-			type Output = $from;
-
-			fn bitxor(self, rhs: Self) -> Self::Output {
-				&self ^ &rhs
-			}
-		}
-
-		impl Default for $from {
-			fn default() -> Self { $from::new() }
-		}
-
-		#[deprecated(
-			since = "0.2.5",
-			note = "will be removed because not big-/little-endian aware"
-		)]
-		impl From<u64> for $from {
-			fn from(mut value: u64) -> $from {
-				let mut ret = $from::new();
-				for i in 0..8 {
-					if i < $size {
-						ret.0[$size - i - 1] = (value & 0xff) as u8;
-						value >>= 8;
-					}
+				if r > 0 {
+					return $crate::core::cmp::Ordering::Greater;
 				}
-				ret
+				$crate::core::cmp::Ordering::Equal
 			}
 		}
 
-		#[deprecated(
-			since = "0.2.5",
-			note = "misses proper error handling; will be replaced by `new_from_slice` in version 0.3"
-		)]
-		impl<'a> From<&'a [u8]> for $from {
-			fn from(s: &'a [u8]) -> $from {
-				$from::from_slice(s)
+		#[cfg(any(not(feature = "libc"), target_os = "unknown"))]
+		impl $crate::core::cmp::PartialEq for $name {
+			#[inline]
+			fn eq(&self, other: &Self) -> bool {
+				self.as_bytes() == other.as_bytes()
 			}
 		}
 
-		impl_std_for_hash!($from, $size);
-		impl_heapsize_for_hash!($from);
-		impl_libc_for_hash!($from, $size);
-		impl_quickcheck_arbitrary_for_hash!($from, $size);
-	}
-}
-
-/// Implements conversion to and from hash types of different sizes. Uses the
-/// last bytes, e.g. `From<H256> for H160` uses bytes 12..32
-/// CAUTION: make sure to call with correct sizes and the bigger type first or
-/// bad things will happen!
-#[macro_export]
-macro_rules! impl_hash_conversions {
-	($a: ident, $a_size: expr, $b: ident, $b_size: expr) => {
-		impl From<$b> for $a {
-			fn from(value: $b) -> $a {
-				debug_assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
-				let mut ret = $a::new();
-				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(&value);
-				ret
+		#[cfg(any(not(feature = "libc"), target_os = "unknown"))]
+		impl $crate::core::cmp::Ord for $name {
+			#[inline]
+			fn cmp(&self, other: &Self) -> $crate::core::cmp::Ordering {
+				self.as_bytes().cmp(other.as_bytes())
 			}
 		}
 
-		impl From<$a> for $b {
-			fn from(value: $a) -> $b {
-				debug_assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
-				let mut ret = $b::new();
-				ret.0.copy_from_slice(&value[($a_size - $b_size)..$a_size]);
+		#[cfg(feature = "rand-support")]
+		impl $crate::rand::distributions::Distribution<$name>
+			for $crate::rand::distributions::Standard
+		{
+			fn sample<R: $crate::rand::Rng + ?Sized>(&self, rng: &mut R) -> $name {
+				let mut ret = $name::zero();
+				for byte in ret.as_bytes_mut().iter_mut() {
+					*byte = rng.gen();
+				}
 				ret
 			}
 		}
 
-		impl<'a> From<&'a $b> for $a {
-			fn from(value: &'a $b) -> $a {
-				let mut ret = $a::new();
-				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(value);
-				ret
+		/// Utilities using the `rand` crate.
+		#[cfg(feature = "rand-support")]
+		impl $name {
```

Comment:
I can understand your critics about expanding the API. I just did the same as I have done already in my crate `apint`. There the `rand`-based API is a good addition and allows for easy testing with the `*_using` APIs and allows for user friendly usage by the APIs with the default RNG. :)

Reply:
Keep in mind that the `rand` API existed earlier but wasn't testable since it was missing the `*_using` methods. So basically the new API allows for greater control and better testability.